subworkflows/local/utils_nfcore_multiplesequencealign_pipeline/main.nf

```diff
@@ -253,96 +257,379 @@ def methodsDescriptionText(mqc_methods_yaml) {
+    def minutes = matcher.group('minutes')?.toDouble() ?: 0.0
+    def seconds = matcher.group('seconds')?.toDouble() ?: 0.0
+    def milliseconds = matcher.group('milliseconds')?.toDouble() ?: 0.0
+
+    return (hours * 60) + minutes + (seconds / 60) + (milliseconds / 60000)
+}
+
+
+/**
+ * Utility function to convert memory to GB.
+ *
+ * This function takes a memory string with units (GB, MB, KB) and converts it to gigabytes (GB).
+ *
+ * Example input formats:
+ * - "16GB"
+ * - "2048MB"
+ * - "1048576KB"
+ *
+ * @param memory The memory string to be converted.
+ * @return The memory in gigabytes as a double, or null if the input is invalid.
+ */
+def convertMemory(String memory) {
+    if (!memory) {
+        return null
+    }
+
+    if (memory.contains("GB")) {
+        return memory.replace("GB", "").toDouble()
+    } else if (memory.contains("MB")) {
+        return memory.replace("MB", "").toDouble() / 1000
+    } else if (memory.contains("KB")) {
+        return memory.replace("KB", "").toDouble() / 1000000
+    }
+    return null
+}
+
+
+/**
+ * Processes the latest trace file in the specified directory.
+ * The trace file is identified based on the given filePattern.
+ *
+ * This function identifies and parses the latest trace file, filters lines related to evaluation,
+ * and converts the trace data into CSV format.
+ *
+ * @param traceDirPath The path to the directory containing trace files.
+ * @param filePattern The pattern to identify the trace files.
+ * @return The parsed CSV data from the trace file.
+ */
+def latesTraceFileToCSV(String traceDirPath, String filePattern) {
+    // Identify and parse the latest trace file based on the given pattern
+    def traceFile = new File(traceDirPath).listFiles().findAll { it.name.startsWith(filePattern) }.sort { -it.lastModified() }.take(1)[0]
+
+    // Keep only the lines that report running times related to evaluation
+    def header = traceFile.readLines()[0].replaceAll("\t", ",")
+    def traceFileAlign = traceFile.readLines().findAll { it.contains("COMPLETED") && it.contains("MULTIPLESEQUENCEALIGN:ALIGN") }.collect { it.replaceAll("\t", ",") }.join("\n")
+    def trace = header + "\n" + traceFileAlign
+
+    // Parse the trace data into CSV format
+    def traceCsv = parseCsv(trace)
+
+    return traceCsv
+}
+
+/**
+ * Merges two lists of maps based on a common ID key.
+ *
+ * @param list1 The first list of maps.
+ * @param list2 The second list of maps.
+ * @param idKey The key used to identify and merge maps from both lists.
+ * @return A new list of merged maps.
+ */
+def mergeListsById(list1, list2, idKey) {
+    // Create a map from list1 with the idKey as the key and the map as the value
+    def map1 = list1.collectEntries { [(it[idKey]): it] }
+
+    // Iterate over list2 and merge with corresponding entries from map1
+    def mergedList = list2.collect { row ->
+        def id = row[idKey]
+        def mergedRow = map1.containsKey(id) ? map1[id] + row : row
```

Comment:
If I am following correctly, this takes into account ids present in map2 and not map1, but it will leave out of the merged list ids present in map1 but not in map2

Reply:
You are correct - it is also what is the expected behaviour but it is rue that is not documented - i have updated the comments :)